### PR TITLE
allow renames() for inlined writes without data-dir

### PR DESCRIPTION
**Pull request by harshavardhana**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request, I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
allow renames() for inlined writes without data-dir

## Motivation and Context
data-dir not being present is okay. However, we can still 
rely on the `rename()` atomic call instead of relying on the 
`xl.meta` write, which may truncate the file incorrectly.

## How to test this PR?
CI/CD should handle it

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by bh4t** [comment]
Can someone please review this?

File: cmd/xl-storage.go [hunk 1]
```diff
@@ -2619,68 +2619,56 @@ func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath string, f
...
-		if !fi.Versioned && !fi.Healing() {
-			// Use https://man7.org/linux/man-pages/man2/rename.2.html if possible on unversioned bucket.
-			if err := Rename2(pathutil.Join(srcVolumeDir, srcPath), pathutil.Join(dstVolumeDir, dstPath)); err == nil {
-				return sign, nil
-			} // if Rename2 is not successful fallback.
-		}
+	if !fi.Versioned && !fi.Healing() {
+		// Use https://man7.org/linux/man-pages/man2/rename.2.html if possible on unversioned bucket.
+		if err := Rename2(pathutil.Join(srcVolumeDir, srcPath), pathutil.Join(dstVolumeDir, dstPath)); err == nil {
+			return sign, nil
```
Comment (klauspost) [inline cmd/xl-storage.go]: Since we are calling directly:
```suggestion
			diskHealthCheckOK(ctx, err)
			return sign, nil
```

**Review by klauspost — APPROVED** [review]
LGTM. Didn't test, though.

**Comment by harshavardhana** [comment]
PTAL @vadmeste @poornas @krisis

File: cmd/xl-storage.go [hunk 2]
```diff
@@ -2623,61 +2623,49 @@ func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath string, f
...
-			}
-			s.deleteFile(dstVolumeDir, dstFilePath, false, false)
-			return 0, err
+	// Commit meta-file
+	if err = renameAll(srcFilePath, dstFilePath, dstVolumeDir); err != nil {
+		if legacyPreserved {
+			// Any failed rename calls un-roll previous transaction.
+			s.deleteFile(dstVolumeDir, legacyDataPath, true, false)
 		}
+		s.deleteFile(dstVolumeDir, dstFilePath, false, false)
```
Comment (vadmeste) [inline cmd/xl-storage.go]: shouldn't this be dstDataPath instead of dstFilePath ?

**Review by vadmeste — CHANGES_REQUESTED** [review]

File: cmd/xl-storage.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/xl-storage.go]: Why? You are renaming dstFilePath, dstDataPath is already done above.

Comment (vadmeste) [reply cmd/xl-storage.go]: > Why? You are renaming dstFilePath, dstDataPath is already done above.

But isn't this xl.meta ? won't we lose old xl.meta content if there is an error with renameAll, such as EIO

Comment (harshavardhana) [reply cmd/xl-storage.go]: It's in temp path @vadmeste not on actual namespace.

Comment (harshavardhana) [reply cmd/xl-storage.go]: I think you are right, we should revisit this.

Comment (harshavardhana) [reply cmd/xl-storage.go]: It must be dstDataPath not dstFilePath @vadmeste updated PTAL.

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**